les/clientpool.go

```diff
@@ -103,296 +106,180 @@ type clientPool struct {
 	}
-	e := &clientInfo{
-		pool:            f,
-		peer:            peer,
-		address:         freeID,
-		queueIndex:      -1,
-		id:              id,
-		connectedAt:     now,
-		priority:        posBalance != 0,
-		posFactors:      f.defaultPosFactors,
-		negFactors:      f.defaultNegFactors,
-		balanceMetaInfo: pb.meta,
+	f.ns.SetState(node, clientFlag, nodestate.Flags{}, 0)
+	f.ns.SetField(node, clientField, c)
+	f.ns.SetField(node, freeIdField, freeID)
+	if c.balance, _ = f.ns.GetField(node, f.BalanceField).(*lps.NodeBalance); c.balance == nil {
+		f.disconnect(peer)
+		return 0, nil
 	}
-	// If the client is a free client, assign with a low free capacity,
-	// Otherwise assign with the given value(priority client)
-	if !e.priority || capacity == 0 {
-		capacity = f.freeClientCap
+	c.balance.SetPriceFactors(f.defaultPosFactors, f.defaultNegFactors)
+	if reqCapacity < f.minCap {
+		reqCapacity = f.minCap
 	}
-	e.capacity = capacity
-
-	// Starts a balance tracker
-	e.balanceTracker.init(f.clock, capacity)
-	e.balanceTracker.setBalance(posBalance, negBalance)
-	e.updatePriceFactors()
-
-	// If the number of clients already connected in the clientpool exceeds its
-	// capacity, evict some clients with lowest priority.
-	//
-	// If the priority of the newly added client is lower than the priority of
-	// all connected clients, the client is rejected.
-	newCapacity := f.connectedCap + capacity
-	newCount := f.connectedQueue.Size() + 1
-	if newCapacity > f.capLimit || newCount > f.connLimit {
-		var (
-			kickList     []*clientInfo
-			kickPriority int64
-		)
-		f.connectedQueue.MultiPop(func(data interface{}, priority int64) bool {
-			c := data.(*clientInfo)
-			kickList = append(kickList, c)
-			kickPriority = priority
-			newCapacity -= c.capacity
-			newCount--
-			return newCapacity > f.capLimit || newCount > f.connLimit
-		})
-		bias := connectedBias
-		if f.disableBias {
-			bias = 0
-		}
-		if newCapacity > f.capLimit || newCount > f.connLimit || (e.balanceTracker.estimatedPriority(now+mclock.AbsTime(bias), false)-kickPriority) > 0 {
-			for _, c := range kickList {
-				f.connectedQueue.Push(c)
-			}
-			clientRejectedMeter.Mark(1)
-			log.Debug("Client rejected", "address", freeID, "id", peerIdToString(id))
-			return false
-		}
-		// accept new client, drop old ones
-		for _, c := range kickList {
-			f.dropClient(c, now, true)
-		}
+	if reqCapacity > f.freeClientCap && c.balance.Priority(now, reqCapacity) <= 0 {
+		f.disconnect(peer)
+		return 0, nil
 	}
 
-	// Register new client to connection queue.
-	f.connectedMap[id] = e
-	f.connectedQueue.Push(e)
-	f.connectedCap += e.capacity
-
-	// If the current client is a paid client, monitor the status of client,
-	// downgrade it to normal client if positive balance is used up.
-	if e.priority {
-		f.priorityConnected += capacity
-		e.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
+	f.ns.SetState(node, f.InactiveFlag, nodestate.Flags{}, 0)
```

Comment:
It's weird. When the client is connected, the balanceTracker will assign it a balance and priority pool will try to tag it a `active` or `inactive` flag. Here the status can already be `active`.

Reply:
`PriorityPool` only cares about nodes that have either `active` or `inactive` flag. Nodes are first switched to `inactive` state externally, then `PriorityPool` can switch between the two states. This behavior of `PriorityPool` is documented already.